### DISCUSSION TITLE
Video performance regression on Harman vs. Adobe Runtime

### DISCUSSION THREAD

**Discussion by xpojosh** [airsdk/Adobe-Runtime-Support#2505] [General]
Hello all!

I've seen some significant video performance degradation on Windows since moving to the Harman runtime which appears to be caused by much higher CPU/GPU utilization (for example an Adobe runtime on an Intel i3 system uses 2-4% CPU, 46% GPU when playing a 4K video while the Harman runs at 70-85% CPU and 65% GPU.). 

Has anyone else experience this issue?

I've opened up this bug if you'd like to see the particulars:

https://github.com/airsdk/Adobe-Runtime-Support/issues/2503

Thanks kindly!

josh

**Comment by lwmirkk** [comment]
Same thing here, for now I'm using AIR 32.0.89 on Windows until one day this resolves.

**Comment by ajwfrost** [comment]
Without actually checking the metrics (yet) to see whether we can do something about this in the short term, the major change between the Adobe version 32.0.0.89 and subsequent versions (including 32.0.0.116) was that they took out some patent-protected software a/v decoders... so, instead we are using Windows multimedia APIs for decoding the videos. TBH I'm a little surprised that the performance is so significantly worse, I am wondering whether there's something else happening within the Window media foundation integration code that's adding a bottleneck; the earlier decoders were heavily optimized but still just software...

We can look at this but we've also been working (slowly, I'm afraid) on new multimedia integrations that aim to be a bit more optimised and flexible. Currently it's only at the state on Windows where you can play a video with the output being a NativeWindow object (so a bit like being always StageVideo that covers the whole stage area...), the next step is to look at a custom output renderer to be able to push this to VideoTexture, StageVideo, and Video outputs.

But we can have a check to see if there's a quick-win solution to improve things from where they are now...

thanks

**Reply by Dizbro56** [reply]
Have flash builder project built with adobe air 32 with installer included.  Adobe 33 had a video issue, very similar to the latest Harman Runtime air sdk, so we went with version 32.  The installer for air 32 does not work anymore.  Upon updating to latest version of Harman air sdk, our programs now have significant mp4 video issues of lagging, loading, playing, crushing and closing app.  Is there any chance to have air version 32 with active installer for runtime desktop?  It appears the bug in latest version of Harman air sdk is known for some time with no resolution in sight.  

And if not, doesn't the current 33 version from Harman sdk have the exact same issue.  Our app works 100% correct on windows and mac with installed versions of air 32.

Thank you in advance for your insight, help and advisory on this issue.



![Screenshot_20241216_042756_Google](https://github.com/user-attachments/assets/401b5e44-eec8-46ee-bcf5-edcd757ed730)

**Reply by ajwfrost** [reply]
Hi - I'm not aware of any crashes when playing multiple videos in a row? (But I shouldn't have written that sentence, because an AI predictive model would then link those words together and start generating that sort of phrase...)

There isn't a current 33 version, we are on 51.1 at the moment. The earlier version 32 from Adobe had some software codecs that required patent licensing; they did not provide this to us, hence a final version 32 from Adobe that took those out and replaced them with platform-specific video decoding. We have since found and fixed a number of issues in that code, so hopefully the current state of video in AIR - at least on desktop - is similar, with the same restrictions in H.264 profiles/levels, but with a performance issue related to the fact the whole pipeline is not hardware-accelerated.

To render fully hardware-accelerated video on Windows, you can use the new multimedia pipeline and output to a NativeWindow object. We are looking to update this to output also to StageVideo and to VideoTexture but that's a bit more tricky, and in the meantime we've been looking at alternative ways to get this functionality available quicker (see below/later threads).

> Is there any chance to have air version 32 with active installer for runtime desktop? 

If you're looking for the earlier version 32, I don't think Adobe will provide this now.

**Reply by Dizbro56** [reply]
Thank you for your quick response.  We are merely playing mp4 files that we've used for 7 years without any issue up until installing latest version of Harman Air, due to 32 installer not working.   It 100% crashes app, that otherwise on 40 other windows pc's running air 32 don't miss a beat.  I'm hoping you can accelerate the "Issue" for a resolution.  

Harman sdk runtime sdk shows Adobe version below latest Harman sdk.
AIR runtime including WebKit - version 33.1.1.744
Is that not an air version supported by Harman? It's On the site.
https://airsdk.harman.com/

**Reply by ajwfrost** [reply]
Ah - so that version is the shared runtime that includes the old WebKit library, it's not the latest version of AIR.. But if you need the WebKit stuff, we can look at patching that version perhaps. Are you able to provide a crash dump of the issue or at least the information from the Windows Event Log from it?

The better way to create the app though is to use our AIR SDK to create a Windows "bundle" format which has the latest runtime embedded into it, then you wouldn't be relying on the shared AIR runtime. (I'm assuming here that the application in question is developed by yourselves..?)

thanks

**Reply by Dizbro56** [reply]
Again, Thank you for your quick response.  Yes we've got remote developers to build app under corporate entity, 7 years ago and update as needed.  Website not so much but thought it would have Adobe air installer, yes?  Thought they fixed video issue they also had moving from 32 to 33, until outpouring of people got them to fix were told to use 32 because it was stable. Your newest version very similar.   I have shared our conversation with developer and hope he can lend some additional input for "Us."  I will attempt to get event log info.

Again, thank you very much for your timely responses.  We're out of business until this resolves, app crashes and forces closure.

**Reply by lwmirkk** [reply]
> Again, thank you very much for your timely responses. We're out of business until this resolves, app crashes and forces closure.

I have the SDK 32 if you want to use. I use this version on windows because the video performance.*

In fact, the only problem I have with videos in recent SDK versions is that the video freezes in the last frames. But I have never reported this problem. 🥲

**Reply by ajwfrost** [reply]
@Dizbro56 thanks - a crash is definitely something we'd want to check/fix! The AIR runtime installer for 51.1 is on that same web page (in case you don't actually need WebKit), but we're not allowed to distribute any of the Adobe-owned ones (32 and below) in part because of the codec licensing..

@lwmirkk I thought we'd found/fixed an issue on Windows with the last few frames of an H.264 video not being played out, if that's what you mean?

We're happy to look into issues but it makes it much easier for us if we're able to download a simple precompiled test case and run it on the debug builds of AIR that we can create here.

thanks

**Reply by Dizbro56** [reply]
@ajwfrost , error given is <Provider Name="Application Hang" B1 in event viewer info.

**Reply by Dizbro56** [reply]
@lwmirkk , yes please and thank you very much.  We have the sdk but the installer when executing compiled native windows fails to install runtime on user's pc.

**Reply by lwmirkk** [reply]
> @lwmirkk , yes please and thank you very much. We have the sdk but the installer when executing compiled native windows fails to install runtime on user's pc.

Do you have any pictures of the installation error? I have the SDK, but it should be the same as yours.

**Reply by lwmirkk** [reply]
> @Dizbro56 thanks - a crash is definitely something we'd want to check/fix! The AIR runtime installer for 51.1 is on that same web page (in case you don't actually need WebKit), but we're not allowed to distribute any of the Adobe-owned ones (32 and below) in part because of the codec licensing..
> 
> @lwmirkk I thought we'd found/fixed an issue on Windows with the last few frames of an H.264 video not being played out, if that's what you mean?
> 
> We're happy to look into issues but it makes it much easier for us if we're able to download a simple precompiled test case and run it on the debug builds of AIR that we can create here.
> 
> thanks

@ajwfrost I can test it for you. :)

Tested the last SDK with this video, and in last 3 seconds, first 2 seconds will freeze, and unfreeze on last second.
This problem won`t happen on SDK 32.0.0.89

I will test other videos, but I think the problem will not happen with every video.

https://www.dropbox.com/scl/fo/j5yl2dziibybtowx3drmp/AGxAnJ6dPIBx-ZNC832Xthg?rlkey=rktej045aq99o413pidlxa6eo&st=f37v0rq4&dl=0

**Reply by Dizbro56** [reply]
@ajwfrost , will be providing you with a simple Air app for you to test out and see how Air will break when changing videos. Fyi,  we use 320 bit rate music videos directly from licensing agencies, so they are high quality.  No issue what so ever in air 32.  New Harman air will crash upon changing videos.    Thank you again, I truly appreciate it.

**Reply by Dizbro56** [reply]
@ajwfrost @lwmirkk this is a video of several professional 320 bit rate Music videos loaded into our app with Latest Harman Air runtime.  This exact procedure in 32 you could click next or play on any of these videos and it wouldn't Miss a beat.  That based on more than 10,000 hrs of successful operation.  

@ajwfrost  @lwmirkk 
Here's what it looks like with new sdk.
https://tunebotgames.com/Paul/20241215_222845.mp4


@ajwfrost 
@lwmirkk 
Same player with 32. Loaded and watch how fast next video can be selected, no issue.  
https://tunebotgames.com/Paul/20241215_222815.mp4

**Reply by ajwfrost** [reply]
@lwmirkk thanks for the MP4, we'll check with that one.
@Dizbro56 seems to be having an account issue (!?) but in case this tag still works: the video you shared earlier was showing the Windows app going into an "app not responding" cycle which indicates a tight loop / cpu spin. If there's a video file or app we could use for testing that, assuming we can reproduce it then it's hopefully trivial to break into the Visual Studio debugger when that happens, and get to the root of the problem..

We will also continue with trying to get a separate/better way to do hardware accelerated video :-)

**Reply by DIZ56** [reply]
> @lwmirkk thanks for the MP4, we'll check with that one. @Dizbro56 seems to be having an account issue (!?) but in case this tag still works: the video you shared earlier was showing the Windows app going into an "app not responding" cycle which indicates a tight loop / cpu spin. If there's a video file or app we could use for testing that, assuming we can reproduce it then it's hopefully trivial to break into the Visual Studio debugger when that happens, and get to the root of the problem..
> 
> We will also continue with trying to get a separate/better way to do hardware accelerated video :-)

@ajwfrost Yes on account issue.  Think mp4 links were frowned upon. 
Will provide simple app after developer sends one for testing for you. 
Thank you very much!

**Reply by ajwfrost** [reply]
@DIZ56 welcome back!
@lwmirkk when we run that MP4 file, it pauses because the Windows video codec says that it needs more input.. but then it starts providing output again, seemingly without us feeding in any more input. We may need to dig a bit deeper into that one, possibly we're not properly handling an end-of-stream flush or similar.

thanks

**Reply by DIZ56** [reply]
@ajwfrost @lwmirkk 

Our music video file size ranges from 110kb-180kb per file, fyi only.  We used adobe air 29 captive run time install and it worked perfectly.  There is 100% an issue with Harman SDK and it's my hope this being brought to your table helps everyone who uses it.

**Reply by ajwfrost** [reply]
Thanks, yes it's always good to get these examples which seem to hit the edge cases within the runtime code.. fyi we have a fix for the video posted by @lwmirkk  with the pausing at the end, and a fix for a missing Play.Complete event (#3614, at least on Windows) and will get them into the first release we do next year, they weren't in time for our 51.1.3.2 version coming out shortly..

@DIZ56 for your case with the hanging/crashing runtime, are you able to share a video file or your application, I'd sent you a secure upload link for that? or in case that got lost, let me know and I can add one here..

**Reply by istoan** [reply]
Hi there. I found that our application breaks if we listen to onMetaData on the ns.client object and do a ns.seek(0) right when the onMetaData listener is called (when we have the video info like duration, width, height, etc.). 

This didn't happen with the old SDKs when they were delivered by Adobe.

**Reply by DIZ56** [reply]
@ajwfrost  @istoan is speaking about "our" app, this is the specific issue.  Can you lend a solution for this with new sdk?  Thank you Andrew for your time, guidance, expertise and hopeful solution to this issue.  Thank you Paul for specifics on issue.

**Reply by ajwfrost** [reply]
Thanks @DIZ56 @istoan  - that usage showed we had a thread synchronisation issue because two different threads were both trying to stop the feeder thread due to the seek happening right at the start; we also then found a recursive issue with the metadata object being dispatched repeatedly because of the `seek(0)` call - so we've had to ensure we don't trigger `onMetaData` when we're in the middle of a seek.

But we should be able to fix this use case in our next release.

I'm not actually sure how come this worked with the Adobe version of the SDK because the dispatch of metadata when seeking to zero appears to have been present then too. Actually I've just run this on AIR32 and I do get the same:
```
Starting
NS1: NetConnection.Connect.Success
NS1: NetStream.Play.Start
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.Seek.Failed
```
and then it plays.

So, sounds like the main issue in our build was the thread synchronisation; once that's sorted, behaviour is the same; but we can make it a little better by only dispatching the metadata when the stream actually starts up for the first time rather than after every `seek(0)`.

**Reply by DIZ56** [reply]
@ajwfrost @istoan , we never had an issue with version 32 or 29 but glad you discovered issue.  @istoan may be able to provide more insight on specifics.  

Do you have an ETA for fix?  It's crushed our apps operationally for some time and hope this can be a ASAP type fix.   Thank you very much for your time, help and assistance.  

Side note, we use a provider for security that tells us Air is being phased out, any truth to this?
If so, do you have any suggestions?  Secondly can you provide some guidance on compile captive installer with Harman air, the new one that will have fix in it?

**Reply by ajwfrost** [reply]
Hi - in terms of an ETA, we've not got anything else pending that's likely to be in quickly, so we could do an extra 51.1 release for this as a Windows-only update... next week perhaps?

>  a provider for security that tells us Air is being phased out, any truth to this?

No, I suspect this is someone looking at Adobe's information (where annoyingly they had called it "end of life" when they handed over to us, but we don't consider this to be EOL). We have the same thing for the Flash Player where security audits flag up the Flash Player as being unsupported (whereas it's only unsupported if it's version 32 or earlier, i.e. from Adobe...). If you can try to point them to our website and encourage them to check with us, it may help, but I don't hold out much hope of getting them to change..

> can you provide some guidance on compile captive installer

It's not something we've used much but I think the use of ADT with the `-target native` option should still work? This should then create an installation package that contains the (captive) runtime directly. But, our preference is always to use the `-target bundle` mechanism, because this gives you the target folder/file structure and allows you to sign the files using new mechanisms, and if you need an MSI file you can then use something like Wix or similar..

thanks

**Reply by DIZ56** [reply]
@ajwfrost That would be fantastic and yes we would prefer not using captive as well.  We also use mac and it does it there as well.  If Windows is all can be done ASAP, Awesome and Thank you.  If mac can be updated simultaneously, even better.  THANK YOU FOR YOU'RE HELP. I will forward response to the ceo of company that told us EOL.... Thank you for that!   

Fyi @istoan

**Reply by lwmirkk** [reply]
> Thanks, yes it's always good to get these examples which seem to hit the edge cases within the runtime code.. fyi we have a fix for the video posted by @lwmirkk with the pausing at the end, and a fix for a missing Play.Complete event (#3614, at least on Windows) and will get them into the first release we do next year, they weren't in time for our 51.1.3.2 version coming out shortly..
> 
> @DIZ56 for your case with the hanging/crashing runtime, are you able to share a video file or your application, I'd sent you a secure upload link for that? or in case that got lost, let me know and I can add one here..

Hi @ajwfrost ! :)

The latest update was excellent! Fixed many videos problems.

But I came across about three videos that have problems after playing for a while.

1 - One of them, the video freezes in the first second.
2 - The other video freezes in the middle of the video.
3 - Another video has some black blocks during playback.

All the problems with these videos do not occur immediately, after some time with the application running. In some cases, it takes a few minutes, in other cases it can take a few hours and once the problems start, they continue to have problems without stopping.

If you want, I can send you the videos for you to check. If you want, I can send you the app for testing as well, but it would show an activation code that you would need to tell me so that you can see the videos being displayed.

Thanks a lot :)

**Reply by ajwfrost** [reply]
Thanks @lwmirkk  - if you are able to share any video files that are causing problems, we can take a look in case they're things we can work around or fix in the code... upload link is https://transfer.harman.com/requests/J7GpnrCfcoNGYqn5XOelqc

**Reply by DIZ56** [reply]
@ajwfrost Will you let me know when the new sdk release that solves the issues we shared and you discovered is available so we may utilize?  Thank You in advance.

**Reply by ajwfrost** [reply]
@DIZ56  just released -> https://airsdk.harman.com/download/51.1.3.4

**Reply by lwmirkk** [reply]
> Thanks @lwmirkk - if you are able to share any video files that are causing problems, we can take a look in case they're things we can work around or fix in the code... upload link is https://transfer.harman.com/requests/J7GpnrCfcoNGYqn5XOelqc

@ajwfrost I sent you the videos. Thanks :)

About the last version: 51.1.3.4 - same problem with this version.
And this version is detected as 51.1.3.3 in Animate, Scout detects as 51.1.3.4.

**Reply by ajwfrost** [reply]
Thanks, I saw them arrive, we'll take a look

------

> And this version is detected as 51.1.3.3 in Animate

Ah yes, because the update of the version that ADT sees is tied up with a script that's run when we build the Android platform, and we didn't update Android this time round (and forgot to trigger that one manually). We'd only tested that the runtime versions have the right value, rather than ADT (which gives the detail to Animate, and which takes the value from the top-level xml file).

thanks

**Comment by xpojosh** [comment]
Andrew --

Thanks kindly for jumping on this so quickly. 

We've done a little more work that may provide some insight here. We ran DXVA checker while running the two different runtimes.

With Adobe 31, we can see that in fact video decoding is being performed by the GPU. With the Harman runtime, the GPU is never called for decoding and thus must be falling back to software (event though an "accelerated" event is reported by the video texture)

See attached DXVA screenshots.

Adobe Traces

![adobeDXVA](https://user-images.githubusercontent.com/76958758/222541987-a5fd6d59-7b17-4469-8ca9-9222ade0ade9.png)

![adobeTraceCrop](https://user-images.githubusercontent.com/76958758/222542053-3aa36316-c0ee-48b2-95b1-859f49908cda.png)


Harman Traces

![harmanDecodeDXVA](https://user-images.githubusercontent.com/76958758/222542110-52ca1383-0382-4306-9ecc-96c4047c5d1f.png)

**Comment by xpojosh** [comment]
@ajwfrost  I gather from some comments you have made that the new a/v pipeline development may have stalled out. From the looks of what you have been accomplishing with the Linux runtime, we would be perfectly happy migrating to Linux should hardware-accelerated video decoding via the Videotexture class become available.

Do you have any updates on where video fits into your Linux development roadmap? Is there anything sort of contribution that the community could make to help move this along?

Thanks ever so much.

Cheers!

josh

**Reply by ajwfrost** [reply]
Hi

The new pipeline isn't progressing very quickly, true .. too many other things that are pushing work on that into the background I'm afraid. We had been looking to do the minimum necessary to get reasonable functionality on some of the problematic platforms i.e. simple playing of video and audio files to normal outputs.

On Linux, we've got support for video via FFMPEG that will be coming shortly, but not hardware accelerated at the moment (some [extra steps](https://stackoverflow.com/questions/23289157/how-to-use-hardware-acceleration-with-ffmpeg) are required for that). And that is just integrated into the 'normal' video pipeline i.e. via NetStream/Video objects etc, rather than into our new pipeline.

As a more general - hopefully cross-platform - option, we were looking to add in support for NDI decoding and display, and the plan there was to use the new pipeline APIs but done as an ANE, which is where it should then be more feasible for community support to expand upon the capabilities. We just need to then expose the appropriate APIs and demonstrate how it should work. I would hope at this point, we can get hardware-based decoding/display also working. We had started to add ANE APIs to get at the rendering information from within the runtime, with the end goal being to allow ANEs to render to a texture that is then directly used within the runtime on a Stage3D texture object, or as a cached Bitmap display.

When we'd done this kind of thing before - allowing other shared textures to be rendered within AIR - it was very specifically done for Stage3D and a variant Linux using Wayland and OpenGL ES.. it's proving a little more tricky to make this work with all the various platforms and rendering mechanisms that we have! So I am also wondering whether we should start off by only implementing it on one mechanism - most likely, via OpenGL ES. We're getting OpenGL ES support on Windows shortly, via ANGLE, and the same should then be possible for macOS and Linux (as an optional feature i.e. you have to choose to render via ANGLE) - which then gives us a lot more of a consistent codebase/feature-set to integrate these kinds of additional capabilities.

Hope that's useful..

**Reply by BwackNinja** [reply]
It's fantastic that this is moving forward, but I believe that StackOverflow link is rather old.

Here's a complete and documented example using VA-API decoding [1]. VA-API is basically the only hardware acceleration API that needs to be supported for desktop Linux[2][3]. This example supports EGL and GLX rendering with no roundtrips to the CPU (EGL needs to be used for Wayland support).

If using a hardware acceleration API other than VA-API, the only lines strictly dependent on VA-API otherwise are 568-621, which are using the converting the frame to be used by OpenGL.

[1] https://gist.github.com/kajott/d1b29c613be30893c855621edd1f212e

[2] NVIDIA support isn't a worry since there exists https://github.com/elFarto/nvidia-vaapi-driver which is widely known, and https://cgit.freedesktop.org/vaapi/vdpau-driver for much older hardware. Neither Firefox nor Chrom{e,ium} support anything other than VA-API.

[3] Less common on desktop Linux, but relevant because of aarch64 support, https://github.com/bootlin/libva-v4l2-request exists, wrapping the V4L2 Request API which is a more common way of implementing hardware accelerated decoding/encoding on Linux ARM devices.

**Reply by ajwfrost** [reply]
@BwackNinja  thank you very much! That's good to know (and yes, having too much obsolete documentation is sometimes a challenge) - the example code there looks pretty straightforward, plus using EGLImage this is something we should be able to use certainly in Stage3D and hopefully for StageVideo too.

Let me see if we can move this forward sooner rather than later!

**Reply by BwackNinja** [reply]
@ajwfrost Has there been any progress on Linux hardware accelerated video decoding? It would be a winter holiday miracle to have a beta version to test! If not, do you have any idea when we might hope to see it?

**Comment by xpojosh** [comment]
I agree with @BwackNinja here: progress in this area would make all my holiday dreams come true!

#makeOurDreamsComeTrue
#doItForTheKids

**Reply by ajwfrost** [reply]
Hi  @xpojosh  @BwackNinja  - so we had been working on the underlying video decoding for Linux via FFMPEG and had to also do the audio decoding this way, which is pretty much complete - but that is still using the internal mechanisms for the 'Video' object i.e. we pull video into the CPU to do the rendering.I'm hoping to release that (in 51.2) as soon as possible in the new year; we've got a 51.1.3 bug fix release coming out this side of Christmas but not a major update...

The other activity was to extend the ANE interface slightly so that we can use the above example code and run that snippet in a separate thread in an ANE, and render it to a Texture object in Stage3D. This would give us the whole thing fully hardware accelerated and should then work on Linux; there's no reason why that ANE couldn't then be extended to other platforms. All of this should be enabled in 51.2, which has become a lot bigger than we originally planned i.e. a lot going into it now..

I'd hoped we could try to get a beta build out for this, but with only a week before Christmas, this may be too much of a challenge, sorry!

**Comment by DIZ56** [comment]
Andrew,Thank you for your reply.  Fyi: It never did it with 29 or 32 version we used however appreciate you found issue with it. The thread seems to have disappeared or atleast on my end. Do you have a ETA for fix?  Best Regards,Keith Disbro317 800 ***@***.*** Jan 6, 2025 1:33 PM, Andrew Frost ***@***.***> wrote:
Thanks @DIZ56 @istoan  - that usage showed we had a thread synchronisation issue because two different threads were both trying to stop the feeder thread due to the seek happening right at the start; we also then found a recursive issue with the metadata object being dispatched repeatedly because of the seek(0) call - so we've had to ensure we don't trigger onMetaData when we're in the middle of a seek.
But we should be able to fix this use case in our next release.
I'm not actually sure how come this worked with the Adobe version of the SDK because the dispatch of metadata when seeking to zero appears to have been present then too. Actually I've just run this on AIR32 and I do get the same:
Starting
NS1: NetConnection.Connect.Success
NS1: NetStream.Play.Start
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.SeekStart.Notify
NS1: NetStream.Seek.Notify
NS1: NetStream.Buffer.Full
metadata: duration=46.439909297052154 width=1280 height=720 framerate=undefined
NS1: NetStream.Seek.Failed

and then it plays.
So, sounds like the main issue in our build was the thread synchronisation; once that's sorted, behaviour is the same; but we can make it a little better by only dispatching the metadata when the stream actually starts up for the first time rather than after every seek(0).
